feat(language-service): provide hover for microsyntax in structural directive

ivanwonder opened the pull request:
In #34564, I only support hover over the directive name. And have a problem, when hovering over `*ngFor="exp"`, it will show `ngForOf` property of `NgForOf` class, this pr will show `(directive) CommonModule.NgForOf: class`. This PR also supports the microsyntax(e.g. `trackBy `, `of`).

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ivanwonder committed:
feat(language-service): provide hover for microsyntax in structural directive

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -20,6 +20,43 @@ export interface SymbolInfo {
+
+  // Find the symbol that contains the position.
+  templateBindings.forEach(tb => {
+    if (!tb.keyIsVar) {
```
Prefer filtering by `keyIsVar` first or bailing out if `keyIsVar` is true: `if (tb.keyIsVar) return;`, because there are a lot of nested ifs.

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -20,6 +20,43 @@ export interface SymbolInfo {
+  // Find the symbol that contains the position.
+  templateBindings.forEach(tb => {
+    if (!tb.keyIsVar) {
+      if (tb.expression && inSpan(valueRelativePosition, tb.expression.ast.span)) {
```
Now that we are on TS 3.7, you can do

```suggestion
      if (inSpan(valueRelativePosition, tb.expression?.ast?.span)) {
```

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -36,14 +73,24 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
-              const expressionOffset = attribute.valueSpan.start.offset;
-              span = offsetSpan(result.span, expressionOffset);
+          if (attribute.name.startsWith('*')) {
+            if (attribute.valueSpan) {
```
This check is done in `getSymbolInMicrosyntax`; please remove it in one place.

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -36,14 +73,24 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
+                symbol = result.symbol;
+                const expressionOffset = attribute.valueSpan.start.offset;
+                span = offsetSpan(result.span, expressionOffset);
+              }
```
Please assign to one `result` variable and perform this procedure simultaneously for both branches rather than duplicating.

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -130,12 +177,25 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
+                if (attribute.name.startsWith('*')) {
+                  compileTypeSummary = directive.directive;
+                  symbol = info.template.query.getTypeSymbol(compileTypeSummary.type.reference);
+                  // Use 'attribute.sourceSpan' instead of directive,
```
```suggestion
                  // Use 'attribute.sourceSpan' instead of the directive's,
```

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -130,12 +177,25 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
+                  symbol = findInputBinding(info, ast.templateName, directive);
+                  span = spanOf(ast);
+                }
+              }
```
What happens if you get rid of this and just do

```typescript
if (!directive) return;

symbol = findInputBinding(info.templateAst, ast, templatePosition);
span = spanOf(ast);
```

The logic here seems very similar to what you have written and changed for `attributeValueSymbol`. Is it necessary if we are already on a directive property?

ivanwonder replied:
For example `*ngFor="exp"`, the `attributeValueSymbol` only checks the `exp`, if it returns false, for the structural directive, we should get the `ngFor` directive, for attribute directives `[model]="value"`, we need the directive property `model` from `string-model`.

ivanwonder commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -36,15 +36,10 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
+            symbol = result.symbol;
+            span = offsetSpan(result.span, attribute.valueSpan !.start.offset);
           }
           return true;
```
rename the function `getSymbolInMicrosyntax` to `getSymbolInAttributeValue`.  the `getExpressionSymbol` have been done in `getSymbolInAttributeValue`, so remove it.

ivanwonder replied:
If it's ok, then I can remove the arguments `ast`, `inEvent` of function `attributeValueSymbol`, they are unused now.

ayazhafiz replied:
Yes, feel free to remove them!

ivanwonder replied:
The arguments `ast`, `inEvent` have been removed.

ayazhafiz commented on packages/language-service/test/definitions_spec.ts:
```diff
@@ -264,29 +264,84 @@ describe('definitions', () => {
+      const def = definitions ![0];
+      expect(def.fileName).toBe(refFileName);
+      expect(def.name).toBe('NgForOf');
+      expect(def.kind).toBe('property');
```
Hm, this doesn't look right.. `NgForOf` is the directive, `ngForOf` is the property on that directive:

https://github.com/angular/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/packages/common/src/directives/ng_for_of.ts#L124-L133

This is why I asked about the code change in `visitDirectiveProperty`; it looks like it is grabbing the directive name rather than the property.

ivanwonder replied:
I think it's the bug here
https://github.com/angular/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/packages/language-service/src/typescript_symbols.ts#L358
The hover info is correct because of the code here.
https://github.com/angular/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/packages/language-service/src/hover.ts#L139

ivanwonder replied:
> This is why I asked about the code change in `visitDirectiveProperty`; it looks like it is grabbing the directive name rather than the property.

Yes, I use the directive name here, see the hover info `(directive) CommonModule.NgForOf: class`.

ivanwonder replied:
If you want to see the `ngForOf` property, hover over the attribute value
`<div *ngFor="let item «ᐱofᐱ heroes»; trackBy: test;"></div>`.
I have added the test for `trackBy`.

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -130,18 +125,67 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
   }
 }
 
+function getSymbolInAttributeValue(
```
Nice! I like this function a lot.

ayazhafiz reviewed (commented):
My apologies for the delayed review. I will try to respond as soon as possible next time..

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -129,19 +124,68 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
+    } else if (inSpan(valueRelativePosition, tb.span)) {
+      const template = path.first(EmbeddedTemplateAst);
+      if (template) {
+        // One element can only have one template bindings.
```
```suggestion
        // One element can only have one template binding.
```

ayazhafiz commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -129,19 +124,68 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
+      const template = path.first(EmbeddedTemplateAst);
+      if (template) {
+        // One element can only have one template bindings.
+        const directiveAst = template.directives[0];
```
What happens if this directive does not have the matching input binding?

ivanwonder replied:
It will be `undefined` as before. Here, firstly I will check the input value, then check the property.
for example `<div *ngFor="let item of heroes; «ᐱtrackByᐱ: test»;">`, check `test`, and then `trackBy`.
For `<div *ngIf="«include»"></div>`, it only check the input value, the `tb.span` wil be `{start:0, end:0}`. So here I only try to get the `EmbeddedTemplateAst`.

ayazhafiz reviewed (changes requested):
lgtm just a couple notes

ayazhafiz reviewed (approved):
thanks!

kyliau commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -129,19 +124,68 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
 }
 
+function getSymbolInAttributeValue(
+    info: AstResult, templatePosition: number, path: TemplateAstPath,
```
`templatePosition = path.position`, so it is not needed as a function parameter

ivanwonder replied:
I think need the `templatePosition` to locate the symbol.

kyliau replied:
Sorry, I meant to say that `templatePosition` is the same as `path.position`, so anywhere `templatePosition` is used in this function, they can be replaced with `path.position`.

ivanwonder replied:
I don’t think the path has property position. Do you mean the path.tail?

ayazhafiz replied:
it does:

https://github.com/angular/angular/blob/32b72f39f061773dfc455446106ad0bbcd5be88c/packages/compiler/src/ast_path.ts#L27

ivanwonder replied:
my mistake.

kyliau commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -129,19 +124,68 @@ export function locateSymbol(info: AstResult, position: number): SymbolInfo|unde
   }
 }
 
+function getSymbolInAttributeValue(
```
Could you please add function doc?

kyliau commented on packages/language-service/test/definitions_spec.ts:
```diff
@@ -264,29 +264,84 @@ describe('definitions', () => {
+      expect(textSpan).toEqual(boundedText);
+
+      expect(definitions).toBeDefined();
+      expect(definitions !.length).toBe(2);
```
Could you please add an assertion for `definitions[1]`? 
Is it the same as `definitions[0]`?
If so, could you please add a comment to explain why this is the case?

ivanwonder replied:
yes, it's the same as the `definitions[0]`. comment has added.

ayazhafiz replied:
Can you add tests for the second definition?

ivanwonder replied:
The two definitions are setter and getter of 'ngForTrackBy'. @ayazhafiz

ivanwonder replied:
I just add a comment, do we need to test the getter and setter?

ayazhafiz replied:
Yes; could you add expectations in this test method for `definitions[1]`? You can loop over all definitions and make sure that they have the same fileName, name , and kind.

ivanwonder replied:
OK, How about removing the `expect(definitions !.length).toBe(2);`, just make sure they have the same fileName, name, and kind.

ayazhafiz replied:
I think the expectation that there is 2 is good, because of the comment that you have added. And we also want to verify that both of the definitions are correct (that their file and name and kind is what you expect)

kyliau commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -193,17 +237,14 @@ function findParentOfBinding(
-      if (classSymbol) {
-        return classSymbol.members().get(fieldName);
-      }
+function findInputBinding(info: AstResult, name: string, directiveAst: DirectiveAst): Symbol|
```
Please add function doc.

kyliau reviewed (changes requested):
Minor changes. Thank you for adding this feature!

ivanwonder commented:
@ayazhafiz I just rebase to the master. can you look at the test case error? are they the expected way? I think it's a bug for `definitions`.

ayazhafiz commented:
The `property` for NgForOf I think is. I have noticed the same thing with a commit I was working on today. I will take a look and fix in a few hours if that is okay. My apologies for the trouble.

ayazhafiz commented:
@ivanwonder I think for now, the easiest and most correct thing to do for this PR will be to overwrite the symbol found for the structural directive in `visitDirectiveProperty` to be a directive symbol, via e.g. `symbol = symbol && new OverrideKindSymbol(symbol, DirectiveKind.DIRECTIVE)`. This doesn't fix the `CommonModule.` prefix missing, but that requires a PR that I don't want to block this feature on. Let me know if that sounds good..

ivanwonder commented:
@ayazhafiz So, I also need to make the assertion work for the length of `definitions`?

ayazhafiz commented:
Yes. I do not think that is a bug.

ivanwonder committed:
fix(language-service): remove repeated symbol definitions for structural directive

For the structural directive, the 'path' will contain multiple `BoundDirectivePropertyAst` which depends on the number of directive property in the attribute value(e.g. '*ngFor="let item of []; trackBy: test;"', it has 2 `BoundDirectivePropertyAst`, 'ngForOf' and 'ngForTrackBy').

ivanwonder commented on packages/language-service/src/locate_symbol.ts:
```diff
@@ -38,6 +40,11 @@ export function locateSymbols(info: AstResult, position: number): SymbolInfo[] {
+  // For the structural directive, only care about the last template AST.
+  if (attribute?.name.startsWith('*')) {
+    toVisit.splice(0, toVisit.length - 1);
+  }
```
For the structural directive, the 'path' will contain multiple `BoundDirectivePropertyAst` which depends on the number of directive property in the attribute value(e.g. '*ngFor="let item of []; trackBy: test;"', it has 2 `BoundDirectivePropertyAst`, 'ngForOf' and 'ngForTrackBy').

ivanwonder replied:
@ayazhafiz I think this is why the length of `definitions` has doubled.

ivanwonder replied:
You can review my PR again. If it’s ok, can you approve it?
I test it in vscode, the number of definitions show in the tooltip doesn’t ignore the duplication.

ivanwonder replied:
![image](https://user-images.githubusercontent.com/15815660/73156797-e4fd1300-4119-11ea-8eed-979b97da0b67.png)
https://github.com/angular/angular/blob/3d4b1431ecbf6e6d6605fea08016baf905fc3299/packages/language-service/test/definitions_spec.ts#L66
When you click it, it will show only one definition.

ivanwonder replied:
Maybe we should filter the `definitions` here.
https://github.com/angular/angular/blob/3d4b1431ecbf6e6d6605fea08016baf905fc3299/packages/language-service/src/definitions.ts#L68-L71

ayazhafiz replied:
Yes, we will filter the definitions. The PR looks good to me. I think for now this is fine, will fixup after the definition-pruning PR.

kyliau replied:
There is a `uniqueBySpan` function in `diagnostics.ts` that could be useful here.
